File: 0055_jump_game.py
# n = len(nums)
# Time: O(n)
# Space: O(1)
from typing import List


class Solution:
    def canJump(self, nums: List[int]) -> bool:
        goal = len(nums) - 1

        for i in range(len(nums) - 2, -1, -1):
            if i + nums[i] >= goal:
                goal = i

        if goal == 0:
            return True

        return False

File: go/0055_jump_game.go
package main

// n = len(nums)
// Time: O(n)
// Space: O(1)
func canJump(nums []int) bool {
	farthest := 0
	for i := 0; i < len(nums); i++ {
		if i > farthest {
			return false
		}

		farthest = max(i+nums[i], farthest)
	}

	return true
}
